### PR TITLE
feat: use /rest/packages when searching by checksum

### PR DESCRIPTION
**Pull request by alexandru-dragomir**
#### What does this PR do?

Change the unmanaged package search to use /rest/packages
instead of maven central. 

#### What are the relevant tickets?

[Jira ticket TARDIS-1031](https://snyksec.atlassian.net/browse/TARDIS-1031)

### CODE REVIEW

File: lib/archive.ts [hunk 1]
```diff
@@ -3,57 +3,36 @@ import * as path from 'path';
 import * as crypto from 'crypto';
 import * as fs from 'fs';
 import * as glob from 'glob';
-import * as needle from 'needle';
```
Comment (gitphill) [inline lib/archive.ts:6]: can we remove needle as a dependency now?

File: lib/index.ts [hunk 2]
```diff
@@ -115,9 +117,17 @@ export async function inspect(
     options = { dev: false, scanAllUnmanaged: false };
   }
 
+  if (!snykHttpClient) {
+    throw new Error('No HTTP client provided!');
```
Comment (gitphill) [inline lib/index.ts:121]: is there any case in which it's ok not to have a HTTP client? like scanning a pom.xml for example
maybe we move this inside the ifs below where a client if definitely required

Comment (gitphill) [reply lib/index.ts:121]: or inside the `createDepGraphFromArchive` function

File: tests/helpers/mock-search.ts [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    const sha1 = requestInfo.qs?.package_sha1 || '';
+    const packageInfo = FIXTURES.get(sha1);
+
+    if (packageInfo) {
+      return packageInfo;
+    }
+  }
+
+  return { res: { statusCode: 404 } as NeedleResponse, body: undefined };
+}
```
Comment (gitphill) [inline tests/helpers/mock-search.ts:88]: i like it 😍

File: lib/archive.ts (hunk 1 above)
Comment (alexandru-dragomir) [reply lib/archive.ts:6]: As far as I can see, the http client returns 

```{
  res: NeedleResponse;
  body: unknown;
}```

and we also use `method: NeedleHttpVerbs;` in the request. Not sure about typescript stuff, but I guess I could just define `HttpResponse { statusCode: int} ` and get do the same for the verb and it should work. Is that what you had in mind?

Comment (gitphill) [reply lib/archive.ts:6]: if we only use the types, can it be moved to a devDependency?

File: lib/index.ts (hunk 2 above)
Comment (alexandru-dragomir) [reply lib/index.ts]: Done.

File: lib/search.ts [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+        debug(
+          `Failed to parse package url components for ${targetPath} using sha1 '${sha1}.`,
+        );
+        return undefined;
+      }
+    })
+    .filter(
+      (mvnPackage: MavenPackage | undefined): mvnPackage is MavenPackage =>
+        mvnPackage !== undefined,
+    );
```
Comment (gitphill) [inline lib/search.ts:92]: does `.filter(Boolean)` achieve the same thing?

File: lib/search.ts [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+  SnykHttpClient,
+} from './parse/types';
+import { PackageURL } from 'packageurl-js';
+import * as debugLib from 'debug';
+import * as path from 'path';
+
+const debug = debugLib('snyk-mvn-plugin');
+
+const PACKAGE_SEARCH_TYPE = 'maven';
+const PACKAGE_SEARCH_ENDPOINT = '/rest/packages';
```
Comment (gitphill) [inline lib/search.ts:14]: we don't need `/rest` here - the http client prefixes that for us

Comment (gitphill) [reply lib/search.ts:14]: see https://github.com/snyk/snyk-gradle-plugin/pull/238/files where we made the same mistake

File: lib/search.ts (hunk 4 above)
Comment (alexandru-dragomir) [reply lib/search.ts:92]: Not directly. `filter(Boolean)` still returns `MavenPackage | undefined` from TypeScript's point of view. What we can do is something like `map { parsePurl() }.filter(Boolean).map { MavenPackage(param?.group || fallback.group ...`

**Review by gitphill — APPROVED** [review]

**Pull request merged**